### PR TITLE
Record and expose last compaction time

### PR DESCRIPTION
**Pull request by rnewson**
For a clustered database, report the earliest compaction time from any
shard. If any shard does not have a recorded compaction time, show
this as null to avoid confusion.

## Checklist

- [ ] Code is written and works correctly;
- [ ] Changes are covered by tests;
- [ ] Documentation reflects the changes;

### CODE REVIEW

**Comment by davisp** [comment]
What's the motivation here? We're already logging when compaction finishes so seems kinda fishy to have it recorded twice. Also not sure on adding it to the API as it'd seem out of place for non-compacting storage engines.

**Comment by rnewson** [comment]
@davisp the motivation is to allow users to know when old revisions are finally gone from the live database files, handy if you need/want to know that a delete has finally happened.

**Comment by davisp** [comment]
Ah makes sense. Only thing that concerns me is that clustered purge work is happening around this as well so we'll want to figure out how best to order things for that.

File: src/couch/src/couch_bt_engine_header.erl [hunk 1]
```diff
@@ -229,6 +236,7 @@ upgrade_disk_version(#db_header{}=Header) ->
         3 -> throw({database_disk_version_error, ?OLD_DISK_VERSION_ERROR});
         4 -> Header#db_header{security_ptr = nil}; % [0.10 - 0.11)
         5 -> Header; % pre 1.2
+        6 -> Header;
```
Comment (rnewson) [inline src/couch/src/couch_bt_engine_header.erl]: @davisp on reflection, I don't think I need to bump disk version as we're only adding a new field.

**Comment by davisp** [comment]
I'm also mulling over how we might make this a bit more general rather than specifically inserting a last_compaction_time or w/e. Seems like having a storage-engine-defined set of extra info would be generally useful. Only sticking point is how we combine that for the cluster endpoint.

Though on a semi-related note I was wanting an option the other day to show me the db info blobs for all shards rather than the combined version. Perhaps if we do that we can make an arbitrary "engine_info" key that allows storage engines to expose any data they might feel like having there?

**Comment by rnewson** [comment]
Hm, does that block this, though? If we make it more generic later, we can re-implement 'last compaction' within that framework. I don't intend to write that framework, do you?

**Comment by davisp** [comment]
@chewbranca Had a PR or diff at least for the separate db info blobs. The PSE level bit of work would be relatively straight forward. Though for the most part on this I'm still worried about the clustered purge interaction and which order we should be thinking of things there.

**Comment by janl** [comment]
@rnewson do you wanna bring this feature forward? I’m asking for a 2.3.0 milestone :)

**Comment by wohali** [comment]
@rnewson ping. If you're not able to get to this in the next week, I'm removing this from the 2.3.0 milestone.

**Comment by wohali** [comment]
This looks pretty gnarly to get back up to speed @rnewson , any chance for 3.x?

**Comment by wohali** [comment]
@rnewson Any plans to get this in? I am preparing to mass-close old PRs that never got merged.

**Comment by wohali** [comment]
```
15:25 <+rnewson> happy to ditch it.
```

**Pull request closed without merging**